CMakeLists.txt

```diff
@@ -264,9 +276,19 @@ if(HARDENING)
       $<$<NOT:$<CONFIG:Debug>>:-U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3>
     )
 
-    try_append_cxx_flags("-Wstack-protector" TARGET hardening)
-    try_append_cxx_flags("-fstack-protector-all" TARGET hardening)
     try_append_cxx_flags("-fcf-protection=full" TARGET hardening)
+    try_append_cxx_flags("-fstack-protector-all" TARGET hardening)
+    try_append_cxx_flags("-Wstack-protector" TARGET hardening)
+
+    try_append_linker_flag("-fstack-protector-all" TARGET hardening)
```

Comment:
Why is this also being added to linker flags?

Reply:
> > It mirrors the master branch behavior:
> 
> Right, but isn't it being used in a compilation context there? What does this flag do as a link flag, and why do we need to test/add it separately?

Libtool passes it through to a linking context explicitly, no?

> > Otherwise, compiling DLL fails.
> 
> What's the failure? It's not clear from that thread.

```
$ cmake -B build --toolchain depends/x86_64-w64-mingw32/share/toolchain.cmake -DBUILD_SHARED_LIBS=ON
$ cmake --build build -t bitcoinconsensus
...
[100%] Linking CXX shared library libbitcoinconsensus.dll
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(cleanse.cpp.obj): in function `memory_cleanse(void*, unsigned long long)':
/home/hebasto/git/bitcoin/src/support/cleanse.cpp:35: undefined reference to `__stack_chk_fail'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(cleanse.cpp.obj):cleanse.cpp:(.rdata$.refptr.__stack_chk_guard[.refptr.__stack_chk_guard]+0x0): undefined reference to `__stack_chk_guard'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(bitcoinconsensus.cpp.obj):/home/hebasto/git/bitcoin/src/script/bitcoinconsensus.cpp:158: undefined reference to `__stack_chk_fail'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(bitcoinconsensus.cpp.obj):/home/hebasto/git/bitcoin/src/script/bitcoinconsensus.cpp:41: undefined reference to `__stack_chk_fail'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(bitcoinconsensus.cpp.obj):/home/hebasto/git/bitcoin/src/serialize.h:391: undefined reference to `__stack_chk_fail'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(bitcoinconsensus.cpp.obj):/home/hebasto/git/bitcoin/src/serialize.h:858: undefined reference to `__stack_chk_fail'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(bitcoinconsensus.cpp.obj):/home/hebasto/git/bitcoin/src/serialize.h:711: undefined reference to `__stack_chk_fail'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(bitcoinconsensus.cpp.obj):/home/hebasto/git/bitcoin/src/serialize.h:711: more undefined references to `__stack_chk_fail' follow
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(hmac_sha256.cpp.obj):/usr/share/mingw-w64/include/string.h:202: undefined reference to `__memcpy_chk'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(hmac_sha256.cpp.obj): in function `CHMAC_SHA256::CHMAC_SHA256(unsigned char const*, unsigned long long)':
/home/hebasto/git/bitcoin/src/crypto/hmac_sha256.cpp:27: undefined reference to `__stack_chk_fail'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(hmac_sha256.cpp.obj): in function `CHMAC_SHA256::Finalize(unsigned char*)':
/home/hebasto/git/bitcoin/src/crypto/hmac_sha256.cpp:34: undefined reference to `__stack_chk_fail'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(hmac_sha512.cpp.obj):/usr/share/mingw-w64/include/string.h:202: undefined reference to `__memcpy_chk'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(hmac_sha512.cpp.obj): in function `CHMAC_SHA512::CHMAC_SHA512(unsigned char const*, unsigned long long)':
/home/hebasto/git/bitcoin/src/crypto/hmac_sha512.cpp:27: undefined reference to `__stack_chk_fail'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(hmac_sha512.cpp.obj): in function `CHMAC_SHA512::Finalize(unsigned char*)':
/home/hebasto/git/bitcoin/src/crypto/hmac_sha512.cpp:34: undefined reference to `__stack_chk_fail'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(poly1305.cpp.obj):/home/hebasto/git/bitcoin/src/crypto/poly1305.cpp:98: undefined reference to `__stack_chk_fail'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(poly1305.cpp.obj): in function `poly1305_donna::poly1305_init(poly1305_donna::poly1305_context*, unsigned char const*)':
/home/hebasto/git/bitcoin/src/crypto/poly1305.cpp:38: undefined reference to `__stack_chk_fail'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(poly1305.cpp.obj): in function `poly1305_donna::poly1305_finish(poly1305_donna::poly1305_context*, unsigned char*)':
/home/hebasto/git/bitcoin/src/crypto/poly1305.cpp:184: undefined reference to `__stack_chk_fail'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(poly1305.cpp.obj):/home/hebasto/git/bitcoin/src/crypto/poly1305.cpp:221: more undefined references to `__stack_chk_fail' follow
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(interpreter.cpp.obj):/usr/share/mingw-w64/include/string.h:202: undefined reference to `__memcpy_chk'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(interpreter.cpp.obj): in function `FindAndDelete(CScript&, CScript const&)':
/home/hebasto/git/bitcoin/src/script/interpreter.cpp:254: undefined reference to `__stack_chk_fail'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(interpreter.cpp.obj):/usr/share/mingw-w64/include/string.h:202: undefined reference to `__memcpy_chk'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(interpreter.cpp.obj): in function `ComputeTapbranchHash(Span<unsigned char const>, Span<unsigned char const>)':
/home/hebasto/git/bitcoin/src/script/interpreter.cpp:1842: undefined reference to `__stack_chk_fail'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(interpreter.cpp.obj): in function `ComputeTaprootMerkleRoot(Span<unsigned char const>, uint256 const&)':
/home/hebasto/git/bitcoin/src/script/interpreter.cpp:1857: undefined reference to `__stack_chk_fail'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(interpreter.cpp.obj): in function `CountWitnessSigOps(CScript const&, CScript const&, CScriptWitness const*, unsigned int)':
/home/hebasto/git/bitcoin/src/script/interpreter.cpp:2122: undefined reference to `__stack_chk_fail'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(interpreter.cpp.obj): in function `CheckSignatureEncoding(std::vector<unsigned char, std::allocator<unsigned char> > const&, unsigned int, ScriptError_t*)':
/home/hebasto/git/bitcoin/src/script/interpreter.cpp:215: undefined reference to `__stack_chk_fail'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(interpreter.cpp.obj):/home/hebasto/git/bitcoin/src/script/interpreter.cpp:404: undefined reference to `__stack_chk_fail'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(interpreter.cpp.obj):/home/hebasto/git/bitcoin/src/script/script.h:379: more undefined references to `__stack_chk_fail' follow
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(interpreter.cpp.obj):/usr/share/mingw-w64/include/string.h:202: undefined reference to `__memcpy_chk'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(interpreter.cpp.obj): in function `GenericTransactionSignatureChecker<CMutableTransaction>::CheckECDSASignature(std::vector<unsigned char, std::allocator<unsigned char> > const&, std::vector<unsigned char, std::allocator<unsigned char> > const&, CScript const&, SigVersion) const':
/home/hebasto/git/bitcoin/src/script/interpreter.cpp:1670: undefined reference to `__stack_chk_fail'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(interpreter.cpp.obj): in function `uint256 SignatureHash<CTransaction>(CScript const&, CTransaction const&, unsigned int, int, long long const&, SigVersion, PrecomputedTransactionData const*)':
/home/hebasto/git/bitcoin/src/script/interpreter.cpp:1633: undefined reference to `__stack_chk_fail'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(interpreter.cpp.obj):/usr/share/mingw-w64/include/string.h:202: undefined reference to `__memcpy_chk'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(interpreter.cpp.obj): in function `GenericTransactionSignatureChecker<CTransaction>::CheckECDSASignature(std::vector<unsigned char, std::allocator<unsigned char> > const&, std::vector<unsigned char, std::allocator<unsigned char> > const&, CScript const&, SigVersion) const':
/home/hebasto/git/bitcoin/src/script/interpreter.cpp:1670: undefined reference to `__stack_chk_fail'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(interpreter.cpp.obj):/home/hebasto/git/bitcoin/src/script/interpreter.cpp:2122: undefined reference to `__stack_chk_fail'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(script.cpp.obj): in function `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string<std::allocator<char> >(char const*, std::allocator<char> const&)':
/usr/lib/gcc/x86_64-w64-mingw32/10-posix/include/c++/bits/basic_string.h:540: undefined reference to `__stack_chk_fail'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(script.cpp.obj): in function `CScriptID::CScriptID(CScript const&)':
/home/hebasto/git/bitcoin/src/script/script.cpp:16: undefined reference to `__stack_chk_fail'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(script.cpp.obj): in function `GetOpName[abi:cxx11](opcodetype)':
/home/hebasto/git/bitcoin/src/script/script.cpp:157: undefined reference to `__stack_chk_fail'
/usr/bin/x86_64-w64-mingw32-ld: CMakeFiles/bitcoinconsensus.dir/objects.a(script.cpp.obj):/home/hebasto/git/bitcoin/src/script/script.cpp:214: more undefined references to `__stack_chk_fail' follow
collect2: error: ld returned 1 exit status
gmake[3]: *** [src/script/CMakeFiles/bitcoinconsensus.dir/build.make:577: src/script/libbitcoinconsensus.dll] Error 1
gmake[2]: *** [CMakeFiles/Makefile2:1631: src/script/CMakeFiles/bitcoinconsensus.dir/all] Error 2
gmake[1]: *** [CMakeFiles/Makefile2:1638: src/script/CMakeFiles/bitcoinconsensus.dir/rule] Error 2
gmake: *** [Makefile:803: bitcoinconsensus] Error 2
```

Reply:
> Libtool passes it through to a linking context explicitly, no?

There's no -Wl, or -X prefix, so how is it getting to the linker?

In any case, that's why my question was, if the linker sees this flag, what happens? From my look at the man page for ld/lld, this isn't an option accepted by any linker, so what does it do?

Reply:
I cannot explain details. The outcome is that linking against `libssp` is not required anymore.

Reply:
> The outcome is that linking against libssp is not required anymore.

I don't really understand how that could be the outcome? The more likely outcome is that by passing `-fstack-protector`, `-lssp` is being added to the link line by the compiler driver, but that doesn't explain why it's not already there, when passing `-fstack-protector` via cxxflags.

> What's the failure? It's not clear from that thread.

This doesn't fail for me. Building this branch at (a281ca282d53ac6ab12d8cbfa101346f035aa6e6), with this diff applied:
```diff
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -280,8 +280,6 @@ if(HARDENING)
     try_append_cxx_flags("-fstack-protector-all" TARGET hardening)
     try_append_cxx_flags("-Wstack-protector" TARGET hardening)
 
-    try_append_linker_flag("-fstack-protector-all" TARGET hardening)
-
     try_append_linker_flag("-Wl,--enable-reloc-section" TARGET hardening)
     try_append_linker_flag("-Wl,--dynamicbase" TARGET hardening)
     try_append_linker_flag("-Wl,--nxcompat" TARGET hardening)
```
using the steps above, works fine. Although I do see warnings about no support for PIE:
```bash
-- The CXX compiler identification is GNU 13.2.0
-- The ASM compiler identification is GNU
-- Found assembler: /opt/homebrew/bin/x86_64-w64-mingw32-gcc
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /opt/homebrew/bin/x86_64-w64-mingw32-g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Warning at CMakeLists.txt:90 (message):
  PIE is not supported at link time: PIE and NO_PIE are not supported by
  linker for CXX
<snip>
[100%] Building CXX object src/script/CMakeFiles/bitcoinconsensus.dir/__/util/strencodings.cpp.obj
[100%] Linking CXX shared library libbitcoinconsensus.dll
[100%] Built target bitcoinconsensus
```

Reply:
Have you tried with GCC 11?

Reply:
No. If it doesn't work for GCC 11, then I guess we should figure out what was broken/what has changed.

Reply:
I think this discussion got carried away. The fact that the current autotools build uses `-fstack-protector-all` when linking is enough justification to have it in CMake build when linking (otherwise somebody will frown CMake does not mimic 1:1 the autotools build). Further, when it is removed, linking fails in some environments, so it is not completely useless.

Reply:
> @hebasto is the plan to add this documentation?

Sure. It'll be done shortly.

Reply:
> is the plan to add this documentation?

A comment has been [added](https://github.com/hebasto/bitcoin/pull/41#issuecomment-1852222107).